Add Resource#allocatedCase

TimWSpence opened the pull request:
Add `Resource#allocatedCase` as a dual of `applyCase`.

Addresses #2472 by tracking ExitCases. Alters existing combinators to take advantage of `allocatedCase` instead of `allocated`

TimWSpence committed:
Add Resource#allocateFull

TimWSpence committed:
Add failing test

TimWSpence committed:
Fix uncancelable

TimWSpence committed:
Implement allocated in terms of allocatedFull

TimWSpence committed:
s/allocated/allocatedFull/g part 1

TimWSpence committed:
s/allocated/allocatedFull/g part 2

TimWSpence committed:
s/allocated/allocatedFull/g part 3

TimWSpence committed:
Add Scaladoc for allocatedFull

TimWSpence committed:
Scalafmt

TimWSpence committed:
Fix Scaladoc

TimWSpence commented:
I haven't ported `start` to use `allocatedFull` yet. It wasn't immediately obvious to me how to do it. I'm assuming it's possible and that I should do it?

TimWSpence committed:
Test that uncancelable is stack-safe

TimWSpence committed:
Failing stack-safety test :(

TimWSpence commented on kernel/shared/src/main/scala/cats/effect/kernel/Resource.scala:
```diff
@@ -383,21 +385,21 @@ sealed abstract class Resource[F[_], +A] {
             resource(poll) flatMap {
               case (b, rel) =>
-                val rel2 = rel(ExitCase.Succeeded).guarantee(release)
+                val rel2 = (ec: ExitCase) => rel(ec).guarantee(release(ec))
```
This is not stack-safe 😭 It would be trivial if we had `Defer` here but I can't force that on things like `uncancelable` that call this now. If anyone has any thoughts then I would love to hear them!

TimWSpence committed:
Fix finalizer stack-safety for allocatedFull

TimWSpence commented on kernel/shared/src/main/scala/cats/effect/kernel/Resource.scala:
```diff
@@ -954,30 +960,32 @@ object Resource extends ResourceFOInstances0 with ResourceHOInstances0 with Reso
                   Kleisli { r =>
-                    nt(rfa.allocated) flatMap { case (a, fin) => r.update(_ !> fin).as(a) }
+                    nt(rfa.allocatedFull) flatMap {
+                      case (a, fin) => r.update(f => (ec: ExitCase) => f(ec) !> fin(ec)).as(a)
```
TODO: This looks stack-unsafe as well. Need to check!

TimWSpence commented on kernel/shared/src/main/scala/cats/effect/kernel/Resource.scala:
```diff
@@ -1078,13 +1086,14 @@ object Resource extends ResourceFOInstances0 with ResourceHOInstances0 with Reso
-            val composedFinalizers = fin !> rfin.allocated.flatMap(_._2)
-            (a, (_: Resource.ExitCase) => composedFinalizers)
+            val composedFinalizers =
+              (ec: ExitCase) => fin(ec) !> rfin(ec).allocatedFull.flatMap(_._2(ec))
```
As above - stack-unsafe?

SystemFw commented:
See if you can add tests for the stack safety issues please. This trick should solve it:
https://github.com/typelevel/cats-effect/blob/series/3.x/kernel/shared/src/main/scala/cats/effect/kernel/Resource.scala#L333

TimWSpence commented:
> See if you can add tests for the stack safety issues please. This trick should solve it: https://github.com/typelevel/cats-effect/blob/series/3.x/kernel/shared/src/main/scala/cats/effect/kernel/Resource.scala#L333

Yeah absolutely! I already added one for `allocatedFull` and thought I'd get the implementation of that sorted first before I looked at the others

And thanks! 👍

TimWSpence committed:
Better stack-safety fix using `F.unit >>` to emualte `defer`

TimWSpence committed:
Failing test for flattenK stack-safety

NPEs but changing 5000 -> 1000 makes it pass?

TimWSpence committed:
Test for allocatedFull stack-safety

TimWSpence committed:
More tinkering

TimWSpence committed:
Fix stack-safety of flattenK

TimWSpence committed:
Explanatory comment

SystemFw commented on kernel/shared/src/main/scala/cats/effect/kernel/Resource.scala:
```diff
@@ -437,9 +440,29 @@ sealed abstract class Resource[F[_], +A] {
+   * library code that needs to modify or move the finalizer for an existing resource.
+   */
+  def allocated[B >: A](implicit F: MonadCancel[F, Throwable]): F[(B, F[Unit])] =
+    allocatedFull.map { case (b, fin) => (b, fin(ExitCase.Succeeded)) }
```
I think this requires `uncancelable`

TimWSpence replied:
🤦 I in fact used this precise example in the `MonadCancel` docs 😂

SystemFw commented on kernel/shared/src/main/scala/cats/effect/kernel/Resource.scala:
```diff
@@ -383,21 +385,22 @@ sealed abstract class Resource[F[_], +A] {
               case (b, rel) =>
-                val rel2 = rel(ExitCase.Succeeded).guarantee(release)
+                // Insert F.unit to emulate defer for stack-safety
+                val rel2 = (ec: ExitCase) => rel(ec).guarantee(F.unit >> release(ec))
```
I suggested this because it solved a very similar problem in Resource, but just double checking: you've fixed an actual failed test with this right?

SystemFw replied:
Ah, I had forgotten the details of our previous convo, you did :)

SystemFw commented:
This is a great addition Tim ❤️

TimWSpence committed:
Allocated needs uncancelable

vasilmkd commented:
Bug fixes for days! Sorry I wasn't around for the review.

armanbilge commented on kernel/shared/src/main/scala/cats/effect/kernel/Resource.scala:
```diff
@@ -373,7 +374,8 @@ sealed abstract class Resource[F[_], +A] {
    * library code that needs to modify or move the finalizer for an existing resource.
    */
-  def allocated[B >: A](implicit F: MonadCancel[F, Throwable]): F[(B, F[Unit])] = {
+  def allocatedFull[B >: A](
```
Bikeshedding question: would `allocatedCase` be a more appropriate name for this method? At least based on the `make` functions, `makeFull` has to do with `Poll`, `makeCase` has to do with tracking the `ExitCase` (as is done here), and `makeCaseFull` has to do with both.

vasilmkd replied:
Yes, I agree with this.

TimWSpence replied:
Thanks! Very good point 👍

TimWSpence committed:
s/allocatedFull/allocatedCase

What was allocatedFull was really a dual of applyCase rather than
applyFull so rename it for consistency

armanbilge commented on kernel/shared/src/main/scala/cats/effect/kernel/Resource.scala:
```diff
@@ -373,7 +374,8 @@ sealed abstract class Resource[F[_], +A] {
    * library code that needs to modify or move the finalizer for an existing resource.
    */
-  def allocated[B >: A](implicit F: MonadCancel[F, Throwable]): F[(B, F[Unit])] = {
+  def allocatedCase[B >: A](
```
```suggestion
  private[effect] def allocatedCase[B >: A](
```

@vasilmkd just want to point out that regardless of our release strategy for patches this time around, we can get this PR in like this by keeping the new API internal since it still fixes a lot of bugs in existing public APIs. Then when we are ready we can un-privatize it :)

vasilmkd replied:
I think we should add it in 3.3.1, but let's wait for Daniel.

armanbilge replied:
@djspiewak 👍 in discord
https://discord.com/channels/632277896739946517/839263556754472990/922206112730722325

armanbilge committed:
Hide `allocatedCase` for now

djspiewak reviewed (approved):
This is awesome thank you!

armanbilge commented:
Failure is b/c of https://github.com/typelevel/cats-effect/pull/2665
```
[info] AsyncPlatform CompletableFuture conversion should
[error]   x cancel CompletableFuture on fiber cancellation
[error]    Expected: java.util.concurrent.CancellationException. Got nothing (AsyncPlatformSpec.scala:40)
```

I think it's just a flaky test, and I think I know how to fix it.
